### ISSUE TITLE
Add support for private accounts

### ISSUE THREAD

**Issue by nex3** [Gargron/mastodon#180]
Following Twitter, private accounts would display toots only to followers (not unauthenticated users or non-followers), and would require follow requests to be authorized by the user in question before they go through. This is something that's really important for marginalized users who want to escape not only active harassment, but the passive ability for hostile people to see into their lives.

I know that this is addressed in [the FAQ](https://github.com/Gargron/mastodon/wiki/FAQ#what-about-private-accounts), but I'm hoping this issue can stand as a place for users to express support for the feature and for progress to be tracked, however long it takes.

**Comment by nex3** [comment]
It looks like GNU social at least notionally supports both of these features within a single, [as described in the FAQ](https://gnusocial.net/doc/faq#faq-20). Even if supporting federated followers of private accounts takes longer, it would be really great to have per-instance privacy in the short term.

**Comment by Gargron** [comment]
I do understand the importance of private accounts. However, I am opposed to solutions that only work within a single instance. It means that all your friends from other servers would have to switch to yours to see your posts, and that runs counter to the idea of federation.

**Comment by nex3** [comment]
It seems to me the fact that privacy is inherently restrictive makes it a little more palatable—you could even imagine someone setting up their own Mastodon node with stringent requirements for registration, and setting their visibility to node-only on purpose.

That said, I do understand your hesitation. What about this as a compromise: support federated privacy controls, but have the UI gracefully degrade across nodes. Specifically, between two nodes:

* Any request for non-public information about a protected user, or actions on that user, results in a 403.
* *In addition* to the above, a request to follow a protected user shows up in their UI as a follower request.
* If the protected user approves the follower request, the prospective follower's requests no longer 403 (even though they aren't following yet).
* If the requester follows and then unfollows the protected account, they lose access to it again.

This requires an extra iteration on the follower's part, but it's properly federated and requires no protocol changes. It could also be cleanly folded into the protocol by simply having the publisher's node send a notification to the subscriber's node saying "your follow request has been approved", which would cause the subscriber's node to re-issue the follow request.

**Comment by Gargron** [comment]
The way subscriptions work (PubSubHubbub) is this:

- Account have public Atom feeds that link to a hub
- You can send a subscription request to the hub for that feed
- When the account publishes something, it notifies the hub
- The hub takes the public feed, takes new items, and delivers them to all subscribers

The hub is independent from the rest of the application, so if a feed is not public, the hub won't be able to deliver it to anyone.

This could be solved by piggybacking "private followers" on the Salmon protocol instead of PubSubHubbub. Salmon is for delivering signed notices to individual accounts (like "alice followed bob" being delivered to bob), but you could technically send private statuses to followers that way. It would, however, at least currently, be incompatible with GNU social. Most importantly, there is no protocol for saying "this feed is private, send a follow request first".

(Quite ironically because there is a follow request verb/activity type in the ActivityStreams protocol iirc)

To sum it up, it's not impossible, but it would take exploring new frontiers rather than simply implementing existing protocols.

**Comment by TazeTSchnitzel** [comment]
I agree with @nex3 that private accounts being restricted to an instance could be a feature and not a bug.

But ideally you could federate it, for sure. Even if only between Mastodon instances (if, say, only Mastodon implemented it).

A concern with federated private accounts is that you'd have to trust each other instance you're letting people follow from. Luckily, you'd actually be given the power to decide this, because it's a private account, so you can accept and reject follow requests as you please.

**Comment by golbette** [comment]
Gonna go ahead and add my 2c to the pile that "make account invisible to other instances" would actually be a nice feature to have and not necessarily a bad one. While it goes against federation as a whole philosophically, its effects would very likely be minimal since a minority of users are going to want to be private in the first place - social networks are inherently social! Such a feature would also serve as a useful 'emergency valve' if someone's personal privacy is being breached.

**Comment by TazeTSchnitzel** [comment]
An emergency valve has its value. A problem with federation is you couldn't really hide and protect your existing posts, but at least your new ones would be safe.

**Comment by Gargron** [comment]
This can be closed now. Separate issue is making private posts work across federation as well, but that's a story for another time. For overview of this update, I'll just link my summary here: https://mastodon.social/users/Gargron/updates/529995

**Comment by caitp** [comment]
Privacy is not just about what you post, but also about who you follow, and who follows you. Those ought to be controllable.
